test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,528 @@
+        };
+
+        uattributes.id = MessageField(Some(Box::new(id)));
+
+        if let Some(ttl) = value.get("ttl").and_then(|ttl| ttl.as_str()) {
+            if let Ok(parsed_ttl) = ttl.parse::<u32>() {
+                uattributes.ttl = parsed_ttl.into();
+            } else {
+                error!("Error: Failed to parse _ttl as u32");
+            }
+        };
+
+        if let Some(permission_level) = value
+            .get("permission_level")
+            .and_then(|permission_level| permission_level.as_str())
+        {
+            if let Ok(parsed_permission_level) = permission_level.parse::<u32>() {
+                uattributes.permission_level = Some(parsed_permission_level);
+            } else {
+                error!("Error: Failed to parse permission_level as u32");
+            }
+        };
+
+        if let Some(commstatus_str) = value.get("commstatus").and_then(|v| v.as_str()) {
+            if let Some(code) = UCode::from_str(commstatus_str) {
+                uattributes.commstatus = Some(code.into());
+            } else {
+                error!("Failed to parse commstatus string into UCode.");
+            }
+        } else {
+            error!("Failed to extract commstatus as a string.");
+        }
+
+        debug!(" uattributes.commstatus: {:?}", uattributes.commstatus);
+
+        if let Ok(reqid) = parse_uuid(&value, "reqid") {
+            uattributes.reqid = MessageField(Some(Box::new(reqid)));
+        } else {
+            let err_msg = "Error parsing UUID reqid: ".to_string();
+            error!("{}", err_msg);
+        }
+
+        if let Some(token) = value.get("token") {
+            if let Some(token_str) = token.as_str() {
+                uattributes.token = Some(token_str.to_owned());
+            } else {
+                error!("Error: token is not a string");
+            }
+        };
+        if let Some(traceparent) = value
+            .get("traceparent")
+            .and_then(|traceparent| traceparent.as_str())
+        {
+            uattributes.traceparent = Some(traceparent.to_owned());
+        } else {
+            error!("Error: traceparent is not a string");
+        };
+
+        // special field
+        let special_fields = SpecialFields::default();
+
+        if special_fields.ne(&SpecialFields::default()) {
+            uattributes.special_fields = special_fields;
+        }
+
+        Ok(WrapperUAttribute(uattributes))
+    }
+}
+
+#[allow(clippy::similar_names)]
+fn parse_uuid(value: &Value, uuid: &str) -> Result<UUID, serde_json::Error> {
+    let get_field = |field: &str| -> Option<u64> {
+        value
+            .get(uuid)
+            .and_then(|uuid| uuid.get(field))
+            .and_then(|uuid| uuid.as_str())
+            .and_then(|value| value.parse().ok())
+    };
+
+    let Some(lsb) = get_field("lsb") else {
+        let err_msg = "Missing ".to_string() + "reqid.lsb field";
+        error!("{}", err_msg);
```

Comment:
Ability to use variables within format string:
```suggestion
        error!("{err_msg}");
```